gnovm/pkg/gnolang/debugger.go

```diff
@@ -0,0 +1,694 @@
+			}
+		case DebugAtRun:
+			switch m.lastDebugCmd {
+			case "si", "stepi":
+				m.DebugState = DebugAtCmd
+				debugLineInfo(m)
+			case "s", "step":
+				if m.DebugLoc != m.PrevDebugLoc {
+					m.DebugState = DebugAtCmd
+					m.PrevDebugLoc = m.DebugLoc
+					debugList(m, "")
+					continue loop
+				}
+			default:
+				for _, b := range m.breakpoints {
+					if b == m.DebugLoc && m.DebugLoc != m.PrevDebugLoc {
+						m.DebugState = DebugAtCmd
+						m.PrevDebugLoc = m.DebugLoc
+						debugList(m, "")
+						continue loop
+					}
+				}
+			}
+			break loop
+		case DebugAtExit:
+			os.Exit(0)
+		}
+	}
+	debugUpdateLocation(m)
+
+	// Keep track of exact locations when performing calls.
+	op := m.Ops[m.NumOps-1]
+	switch op {
+	case OpCall:
+		m.debugCall = append(m.debugCall, m.DebugLoc)
+	case OpReturn, OpReturnFromBlock:
+		m.debugCall = m.debugCall[:len(m.debugCall)-1]
+	}
+}
+
+// debugCmd processes a debugger REPL command. It displays a prompt, then
+// reads and parses a command from the debugger input stream, then executes
+// the corresponding function or returns an error.
+// If the command is empty, the last non-empty command is repeated.
+func debugCmd(m *Machine) error {
+	var cmd, arg string
+	fmt.Fprint(m.DebugOut, "dbg> ")
+	if !m.DebugScanner.Scan() {
+		return debugDetach(m, arg) // Clean close of debugger, the target program resumes.
+	}
+	line := m.DebugScanner.Text()
+	n, _ := fmt.Sscan(line, &cmd, &arg)
+	if n == 0 {
+		if m.lastDebugCmd == "" {
+			return nil
+		}
+		cmd, arg = m.lastDebugCmd, m.lastDebugArg
+	} else if cmd[0] == '#' {
+		return nil
+	}
+	c, ok := debugCmds[cmd]
+	if !ok {
+		return errors.New("command not available: " + cmd)
+	}
+	m.lastDebugCmd, m.lastDebugArg = cmd, arg
+	return c.debugFunc(m, arg)
+}
+
+// initDebugIO initializes the debugger standard input and output streams.
+// If no debug address was specified at program start, the debugger will inherit its
+// standard input and output from the process, which will be shared with the target program.
+// If the debug address was specified, the program will be blocked until
+// a client connection is established. The debugger will use this connection and
+// not affecting the target program's.
+// An error during connection setting will result in program panic.
+func initDebugIO(m *Machine) {
+	if m.DebugAddr != "" {
+		l, err := net.Listen("tcp", m.DebugAddr)
+		if err != nil {
+			panic(err)
+		}
+		print("Waiting for debugger client to connect at ", m.DebugAddr)
+		conn, err := l.Accept()
+		if err != nil {
+			panic(err)
+		}
+		println(" connected!")
+		m.DebugIn = conn
+		m.DebugOut = conn
+	}
+	m.DebugScanner = bufio.NewScanner(m.DebugIn)
+}
+
+// debugUpdateLocation computes the source code location for the current VM state.
+// The result is stored in Debugger.DebugLoc.
+func debugUpdateLocation(m *Machine) {
+	loc := m.LastBlock().Source.GetLocation()
+
+	// File must have an unambiguous absolute path.
```

Comment:
I'm not super familiar with how locations are computed in GnoVM. For the purpose of debugging, I just think that representing internally in the debugger all locations in their canonical form (full pathname + line) is the right thing to do, even if we manipulate them in their relative form in the debugger commands or elsewhere.

Reply:
Yeah, but it is a wrong assumption.

Consider the following code:

```go
package main

import "math"

func main() {
	x := math.Sqrt(144)
	println(x)
}
```

The debugger will work fine until it goes in the source for `Sqrt`. Then it will fail:

```
dbg> l
> math.Sqrt() /home/howl/oc/gno2/sqrt.gno:94
Command failed: open /home/howl/oc/gno2/sqrt.gno: no such file or directory
```

Note the path is wrong, because `loc.File` contains only the name, and to get back to the source ideally we'd trace back the MemPackage (possibly through the Store).

I don't mean to say that this should be fixed now, but `injectLocOnPanic` is doing a better job at basically the same thing. If your version does have improvements on detecting the location, I suggest you merge in the changes